File: matlab/+mestra/+internal/Compare.m
classdef Compare
%Compare  The structural equality rule of specification section 30.
%
%   Byte identity across HDF5 versions is not required and must not be
%   tested, because the library decides the superblock and the object
%   header layout.  The normative comparison between two files that
%   should be the same is
%
%     * the same set of object paths;
%     * at each path, the same kind, group or dataset;
%     * for each dataset, the same dtype including byte order,
%       character set and padding, the same shape, the same maximum
%       shape, the same chunk shape, the same filters with the same
%       parameters, and element-by-element equal contents, with floats
%       compared as bits so that NaN equals NaN;
%     * at each path, the same set of attribute names excluding the
%       machinery names of section 18, and for each, the same dtype
%       and the same value;
%     * the same dimension scale attached to each axis of each
%       dataset, compared by the dimension's name.
%
%   See also mestra.write.

    methods (Static)

        function differences = structural(pathA, pathB)
        %structural  A cell array of differences; empty means equal.
            a = mestra.internal.Compare.snapshot(pathA);
            b = mestra.internal.Compare.snapshot(pathB);
            differences = {};
            ka = a.keys();
            kb = b.keys();
            for i = 1:numel(ka)
                if ~b.isKey(ka{i})
                    differences{end + 1} = sprintf( ...
                        '%s is only in the first file', ka{i}); %#ok<AGROW>
                end
            end
            for i = 1:numel(kb)
                if ~a.isKey(kb{i})
                    differences{end + 1} = sprintf( ...
                        '%s is only in the second file', kb{i}); %#ok<AGROW>
                end
            end
            for i = 1:numel(ka)
                if ~b.isKey(ka{i}), continue, end
                differences = [differences ...
                    mestra.internal.Compare.compareOne(ka{i}, a(ka{i}), ...
                                                       b(ka{i}))]; %#ok<AGROW>
            end
        end

        function out = compareOne(path, x, y)
            out = {};
            if ~strcmp(x.kind, y.kind)
                out{end + 1} = sprintf('%s is a %s and a %s', path, ...
                                       x.kind, y.kind);
                return
            end
            out = [out mestra.internal.Compare.compareAttrs(path, ...
                       x.attrs, y.attrs)];
            if ~strcmp(x.kind, 'dataset'), return, end
            fields = {'type', 'dims', 'maxdims', 'chunk', 'filters', ...
                      'strSize', 'cset', 'strpad', 'scales', 'bits'};
            for i = 1:numel(fields)
                f = fields{i};
                if ~isequal(x.(f), y.(f))
                    out{end + 1} = ...
                        sprintf('%s: %s differs', path, f); %#ok<AGROW>
                end
            end
        end

        function out = compareAttrs(path, x, y)
            out = {};
            names = union(x.keys(), y.keys());
            for i = 1:numel(names)
                n = names{i};
                if ~x.isKey(n)
                    out{end + 1} = sprintf( ...
                        '%s: attribute %s is only in the second file', ...
                        path, n); %#ok<AGROW>
                elseif ~y.isKey(n)
                    out{end + 1} = sprintf( ...
                        '%s: attribute %s is only in the first file', ...
                        path, n); %#ok<AGROW>
                elseif ~isequal(x(n), y(n))
                    out{end + 1} = sprintf('%s: attribute %s differs', ...
                                           path, n); %#ok<AGROW>
                end
            end
        end

        function map = snapshot(path)
        %snapshot  Every object in a file, in the form compared above.
            map = containers.Map('KeyType', 'char', 'ValueType', 'any');
            fid = H5F.open(path, 'H5F_ACC_RDONLY', 'H5P_DEFAULT');
            closer = onCleanup(@() H5F.close(fid)); %#ok<NASGU>
            closePass = mestra.internal.H5.pass(); %#ok<NASGU>
            root = H5G.open(fid, '/');
            scales = mestra.internal.H5.scaleMap(fid);
            mestra.internal.Compare.walk(map, root, '/', 0, scales);
            H5G.close(root);
        end

        function walk(map, gid, prefix, depth, scales)
        %walk  Every object under a group, to a bounded depth.
            H5 = mestra.internal.H5;
            if nargin < 4, depth = 0; end
            if nargin < 5, scales = []; end
            if depth > mestra.internal.Limits.get('maxDepth')
                map([prefix ' (not followed)']) = struct('kind', 'limit', ...
                    'attrs', containers.Map('KeyType', 'char', ...
                                            'ValueType', 'any'));
                return
            end
            rec.kind = 'group';
            rec.attrs = mestra.internal.Compare.attrs(gid);
            map(prefix) = rec;
            for name = H5.children(gid)
                if strcmp(prefix, '/')
                    path = ['/' name{1}];
                else
                    path = [prefix '/' name{1}];
                end
                kind = H5.childType(gid, name{1});
                if strcmp(kind, 'group')
                    sub = H5G.open(gid, name{1});
                    mestra.internal.Compare.walk(map, sub, path, ...
                                                 depth + 1, scales);
                    H5G.close(sub);
                elseif strcmp(kind, 'dataset')
                    did = H5D.open(gid, name{1});
                    map(path) = mestra.internal.Compare.dataset(did, scales);
                    H5D.close(did);
                else
                    map(path) = struct('kind', kind, ...
                        'attrs', containers.Map('KeyType', 'char', ...
                                                'ValueType', 'any'));
                end
            end
        end

        function rec = dataset(did, scales)
            H5 = mestra.internal.H5;
            if nargin < 2, scales = []; end
            info = H5.dsetInfo(did);
            rec.kind = 'dataset';
            rec.attrs = mestra.internal.Compare.attrs(did);
            rec.type = info.type;
            rec.dims = info.dims;
            rec.maxdims = info.maxdims;
            rec.chunk = info.chunk;
            rec.filters = info.filters;
            rec.strSize = -1;
            rec.cset = info.cset;
            rec.strpad = info.strpad;
            rec.scales = cell(1, numel(info.dims));
            for axis = 1:numel(info.dims)
                found = H5.scaleNames(did, axis - 1, scales);
                rec.scales{axis} = sort({found.name});
            end
            if info.isScale
                % A dimension scale holds no value anyone reads, so
                % there is nothing to compare but its shape and type.
                rec.bits = [];
                return
            end
            if strcmp(info.type, 'string')
                rec.strSize = info.strSize;
                rec.bits = H5.readRawStrings(did, info);
            else
                data = H5.readData(did, info);
                rec.bits = typecast(H5.cast(info.type, data(:))', 'uint8');
            end
        end

        function map = attrs(oid)
            H5 = mestra.internal.H5;
            map = containers.Map('KeyType', 'char', 'ValueType', 'any');
            for name = H5.publicAttrNames(oid)
                info = H5.attrInfo(oid, name{1});
                if strcmp(info.type, 'string')
                    map(name{1}) = {info.type, info.size, info.cset, ...
                                    info.strpad, ...
                                    H5.readRawStrAttr(oid, name{1})};
                else
                    value = H5.readAttr(oid, name{1});
                    map(name{1}) = {info.type, ...
                        typecast(H5.cast(info.type, value(:))', 'uint8')};
                end
            end
        end
    end
end

File: julia/src/compare.jl
# Structural equality, the normative comparison of section 30.
#
# Byte identity across HDF5 versions is not required and must not be
# tested, because the library decides the superblock and the object
# header layout.  What two files that should be the same must share is
# below, and nothing else.

"""
    structural_diff(a, b) -> Vector{String}

Every way in which two files differ under the rule of section 30: the
same set of object paths; at each path the same kind; for a dataset
the same dtype including byte order, character set and padding, the
same shape, maximum shape, chunk shape and filters, and equal contents
compared as bytes so that NaN equals NaN; at each path the same
attributes excluding the machinery names of section 18; and the same
dimension scale attached to each axis, compared by the scale's name.

An empty result means the two files are structurally equal.
"""
function structural_diff(a::AbstractString, b::AbstractString)
    out = String[]
    HDF5.h5open(String(a), "r") do fa
        HDF5.h5open(String(b), "r") do fb
            ia = ScaleIndex(fa)
            ib = ScaleIndex(fb)
            pa = object_paths(fa)
            pb = object_paths(fb)
            for p in setdiff(keys(pa), keys(pb))
                push!(out, "only in $(a): $(p)")
            end
            for p in setdiff(keys(pb), keys(pa))
                push!(out, "only in $(b): $(p)")
            end
            for p in sort(collect(intersect(keys(pa), keys(pb))))
                pa[p] == pb[p] ||
                    (push!(out, "$(p): $(pa[p]) here, $(pb[p]) there");
                     continue)
                oa = p == "/" ? fa : fa[p]
                ob = p == "/" ? fb : fb[p]
                compare_attrs!(out, p, oa, ob)
                oa isa HDF5.Dataset || continue
                compare_dataset!(out, p, oa, ob, ia, ib)
            end
        end
    end
    return out
end

"""
    structurally_equal(a, b) -> Bool

True when `structural_diff` finds nothing, which is what section 30
means by two files being the same.
"""
structurally_equal(a, b) = isempty(structural_diff(a, b))

function object_paths(f::HDF5.File)
    out = Dict{String,Symbol}("/" => :group)
    walk_objects(f) do path, obj
        out[path] = obj isa HDF5.Group ? :group : :dataset
    end
    return out
end

function compare_attrs!(out, path, oa, ob)
    na = sort([n for n in keys(HDF5.attributes(oa))
               if !(n in MACHINERY_ATTRS)])
    nb = sort([n for n in keys(HDF5.attributes(ob))
               if !(n in MACHINERY_ATTRS)])
    na == nb || (push!(out, "$(path): attributes $(na) against $(nb)");
                 return)
    for n in na
        aa = read_raw_attr(oa, n)
        ab = read_raw_attr(ob, n)
        aa.ti == ab.ti ||
            push!(out, "$(path)@$(n): dtype $(aa.ti) against $(ab.ti)")
        aa.raw == ab.raw ||
            push!(out, "$(path)@$(n): value $(aa.value) against $(ab.value)")
    end
    return out
end

function compare_dataset!(out, path, da, db, ia, ib)
    ta = type_info(HDF5.datatype(da))
    tb = type_info(HDF5.datatype(db))
    ta == tb || push!(out, "$(path): dtype $(ta) against $(tb)")
    sa, ma = disk_shape(da)
    sb, mb = disk_shape(db)
    sa == sb || push!(out, "$(path): shape $(sa) against $(sb)")
    ma == mb || push!(out, "$(path): max shape $(ma) against $(mb)")
    la, ca, fa_ = dataset_layout(da)
    lb, cb, fb_ = dataset_layout(db)
    ca == cb || push!(out, "$(path): chunk $(ca) against $(cb)")
    la == lb || push!(out, "$(path): layout $(la) against $(lb)")
    sort(fa_) == sort(fb_) ||
        push!(out, "$(path): filters $(fa_) against $(fb_)")
    if !is_scale(da) && sa == sb
        _, ra, _ = read_raw_dataset(da)
        _, rb, _ = read_raw_dataset(db)
        ra == rb || push!(out, "$(path): contents differ")
    end
    axa = axis_scales(da, ia)
    axb = axis_scales(db, ib)
    for axis in 0:(length(sa) - 1)
        nsa = axis + 1 <= length(axa) ? axa[axis + 1][1] : 0
        nsb = axis + 1 <= length(axb) ? axb[axis + 1][1] : 0
        if nsa != nsb
            push!(out, "$(path): axis $(axis) has $(nsa) scales against " *
                       "$(nsb)")
            continue
        end
        nsa == 1 || continue
        va = axa[axis + 1][2]
        vb = axb[axis + 1][2]
        va == vb || push!(out, "$(path): axis $(axis) scale $(va) against " *
                               "$(vb)")
    end
    return out
end
